### PR TITLE
feat: after_begin and before_commit migration callbacks

### PR DESCRIPTION
**Pull request by zachdaniel**
This should cover the use case discussed in #31. For more information, see the issue.

I have added tests, but I still need to add documentation which I will do this evening. Using these callbacks, you can write standard migration code. However, I'm struggling to figure out how to make working with up/down play nicely. Given this migration:

```
defmodule MyMigration do
  ...

  def after_begin() do
    execute "some_sql"
  end

  def change() do
    ...
  end
end
```

If you roll it back, you will get an error that `execute "some_sql"` cannot be rolled back. It seems to me like we need some way to differentiate between up and down. Either `after_begin_up` and `after_begin_down`, or `after_begin(:up)`? I'm unsure.

### CODE REVIEW

File: lib/ecto/migration.ex [hunk 1]
```diff
@@ -278,6 +281,12 @@ defmodule Ecto.Migration do
       import Ecto.Migration
       @disable_ddl_transaction false
       @before_compile Ecto.Migration
+
+      def after_begin(), do: :ok
+
+      def before_commit(), do: :ok
+
+      defoverridable after_begin: 0, before_commit: 0
```
Comment (josevalim) [inline lib/ecto/migration.ex]: The recommendation in Elixir is to tag them as optional callbacks and do not provide default implementations in favor of `function_exported?` checks.

File: lib/ecto/migration/runner.ex [hunk 2]
```diff
@@ -23,7 +23,9 @@ defmodule Ecto.Migration.Runner do
     metadata(runner, opts)
 
     log(level, "== Running #{version} #{inspect module}.#{operation}/0 #{direction}")
-    {time1, _} = :timer.tc(module, operation, [])
+    {time1, _} = :timer.tc(fn ->
+      perform_operation(module, operation)
```
Comment (josevalim) [inline lib/ecto/migration/runner.ex]: I would prefer to move this to `migrator` module, as that's the one that starts the transaction OR we move the logic that starts the transaction here. Otherwise the logic can easily get out of sync. For example, we should also check for `repo.__adapter__.supports_dll_transaction?` below, but we don't. :)

File: lib/ecto/migration/runner.ex [hunk 3]
```diff
@@ -264,6 +266,19 @@ defmodule Ecto.Migration.Runner do
...
 
+  defp perform_operation(module, operation) do
+    if module.__migration__[:disable_ddl_transaction] do
+      apply(module, operation, [])
+    else
+      try do
+        module.after_begin()
+        apply(module, operation, [])
+      after
+        module.before_commit()
```
Comment (josevalim) [inline lib/ecto/migration/runner.ex]: We don't need a try/catch here because `before_commit` should only run if we are going to commit... and we are not going to commit if there is an error. :)

File: test/ecto/migrator_test.exs [hunk 4]
```diff
@@ -431,4 +465,35 @@ defmodule Ecto.MigratorTest do
...
+
+      log = capture_log(fn ->
+        assert down(TestRepo, 10, MigrationWithCallbacks) == :ok
+      end)
+
+      assert log =~ "after_begin_down"
+      assert log =~ "before_commit_down"
+    end
+
+    test "they are not run when the migration is disabled" do
```
Comment (josevalim) [inline test/ecto/migrator_test.exs]: I would remove the "they" from the title as otherwise the description would be "migration callbacks they are not run ...".

File: lib/ecto/migration/runner.ex (hunk 2 above)
Comment (zachdaniel) [reply lib/ecto/migration/runner.ex]: Alright, so I've moved the calls into a sort of naive place, and they result in the error I mentioned. I don't see a great way around resolving that without moving the call to the callbacks into the `attempt/6` code. However, I'm unsure if it is correct to move the transaction start/end into that same place. Where do you think would be best?

Comment (josevalim) [reply lib/ecto/migration/runner.ex]: @zachdaniel On one hand, we don't need to make `execute` available in those callbacks. But that means we should probably pass the `repo` as argument.

Another option is to move where they were and use `repo.in_transaction?` to check if you are inside a transaction before running the callbacks?

Comment (zachdaniel) [reply lib/ecto/migration/runner.ex]: Hmm....Yeah, let’s just check if they are in a transaction with the repo :) good idea.

**Comment by zachdaniel** [comment]
I’ve got one small change to add (discussed in comments) and then some docs. I’ll do those both tomorrow morning and then this should be good to go.

File: test/ecto/migrator_test.exs [hunk 5]
```diff
@@ -431,4 +465,35 @@ defmodule Ecto.MigratorTest do
...
+
+      log = capture_log(fn ->
+        assert down(TestRepo, 10, MigrationWithCallbacks) == :ok
+      end)
+
+      assert log =~ "after_begin_down"
+      assert log =~ "before_commit_down"
+    end
+
+    test "are not run when the migration is disabled" do
```
Comment (samphilipd) [inline test/ecto/migrator_test.exs]: Do you mean when transactions are disabled?

File: lib/ecto/migrator.ex [hunk 6]
```diff
@@ -190,7 +190,19 @@ defmodule Ecto.Migrator do
     else
       {:ok, result} =
         repo.transaction(
-          fn -> send_and_receive(parent, ref, fun.()) end,
+          fn ->
+            if function_exported?(module, :after_begin, 0) do
+              module.after_begin()
```
Comment (samphilipd) [inline lib/ecto/migrator.ex]: This will definitely execute in the context of the migration transaction, right?

Comment (josevalim) [reply lib/ecto/migrator.ex]: Yes.

File: test/ecto/migrator_test.exs (hunk 5 above)
Comment (zachdaniel) [reply test/ecto/migrator_test.exs]: yep :)

File: lib/ecto/migration/runner.ex (hunk 2 above)
Comment (zachdaniel) [reply lib/ecto/migration/runner.ex]: @josevalim so checking if the repo is in a transaction presents a slight issue, which is that adapters raise a `RuntimeError: not implemented` when they don't implement `in_transaction?`. The only way I can think to get around that is to do (maybe with better matching):

```
in_transaction? =
  try do
    repo.in_transaction?()
  rescue
   RuntimeError -> false
  end
```

Comment (zachdaniel) [reply lib/ecto/migration/runner.ex]: That actually causes problems for tests, unfortunately. Without the test adapter being able to determine if its in a transaction, I can't really test this behavior without mocking/stubbing that functionality out somehow. Unless we maintain some corresponding kind of state in the process dictionary, but I have a feeling that could have some interesting repercussions.

**Comment by josevalim** [comment]
@zachdaniel if that is the test repository then we can change the test repository to implement it accordingly.

**Comment by josevalim** [comment]
A simple implementation would use the process dictionary. See the test_repo.exs in the original Ecto repository.

**Comment by zachdaniel** [comment]
Wouldn't that cause problems with something like this?

```
Repo.transaction(fn -> 
  Task.async(fn -> 
    Repo.in_transaction?()
  end) |> Task.await()
end)
```

**Comment by zachdaniel** [comment]
@josevalim alright, I think this is ready now :D Thanks for your help!

File: lib/ecto/migrator.ex [hunk 7]
```diff
@@ -190,7 +190,9 @@ defmodule Ecto.Migrator do
     else
       {:ok, result} =
         repo.transaction(
-          fn -> send_and_receive(parent, ref, fun.()) end,
+          fn ->
+            send_and_receive(parent, ref, fun.())
+          end,
```
Comment (josevalim) [inline lib/ecto/migrator.ex]: Let's revert the changes to this file. :)

File: lib/ecto/migration/runner.ex [hunk 8]
```diff
@@ -264,6 +266,32 @@ defmodule Ecto.Migration.Runner do
...
   ## Helpers
 
+  defp perform_operation(repo, module, operation) do
+    # TODO: Test adapter can never trigger this case, so how do we test it?
+    in_transaction? =
+      try do
+        repo.in_transaction?()
+      rescue
+        _ -> false
+      end
```
Comment (josevalim) [inline lib/ecto/migration/runner.ex]: So we shouldn't need the try/catch if we implement `in_transaction?` in the test adapter. `if function_exported?(repo, :in_transaction?, 0) and repo.in_transaction?()` below should be all we need. :)

Comment (zachdaniel) [reply lib/ecto/migration/runner.ex]: Ah, yeah sorry that slipped my mind. Sorry!

**Comment by josevalim** [comment]
@zachdaniel we just need docs for after_begin/before_commit callbacks and we should be good to go!

**Comment by zachdaniel** [comment]
I really spaced on this, thinking it was ready! I’ll finish that up when I get home in about two hours. Thanks for your patience!

**Comment by zachdaniel** [comment]
@josevalim Let me know what you think of the docs I added!

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by josevalim** [comment]
@samphilipd please give it a try and let us know how it does so we can cut a new release.

**Comment by samphilipd** [comment]
@josevalim this is awesome thanks! Just trying to test it, I get this error:

```
Dependencies have diverged:
* ecto_sql (https://github.com/elixir-ecto/ecto_sql.git)
  different specs were given for the ecto_sql app:

  > In apps/nest_db/mix.exs:
    {:ecto_sql, [env: :prod, git: "https://github.com/elixir-ecto/ecto_sql.git", ref: "master", override: true, manager: :mix]}

  > In deps/ex_machina/mix.exs:
    {:ecto_sql, "~> 3.0", [only: [:dev, :test], env: :prod, hex: "ecto_sql", repo: "hexpm", optional: true]}

  Ensure they match or specify one of the above in your deps and set "override: true"
** (Mix) Can't continue due to errors on dependencies
```

I have specified `override: true` but it doesn't appear to have helped... any ideas?

**Comment by josevalim** [comment]
@samphilipd is deps/ex_machina a dependency of nest_db or another app? If so you need to to make sure that you specify the git dep in the app that depends on ex_machina too.

**Comment by samphilipd** [comment]
Yes, it works! Exactly what I wanted!

I have created our own module `NestDB.migration` so we can `use NestDB.Migration` instead of `use Ecto.Migration`. It is however a bit error-prone to have to remember this every time. Is it possible to modify the template for generated migrations?

**Comment by zachdaniel** [comment]
I think that would be relatively easy. We would just have a configuration on the repo perhaps?

```
config :my_app, MyApp.Repo,
  migration_implementation: MyApp.Migration
```

Or something along those lines.

**Comment by josevalim** [comment]
Although if you are using Phoenix, that would be a Phoenix feature.


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by samphilipd** [comment]
`ecto.gen.migration` is in Phoenix?? That is surprising.

**Comment by josevalim** [comment]
Oh, wait. `ecto.gen.migration` is in Ecto. :) but Phoenix also generates migrations.,

**Comment by samphilipd** [comment]
@zachdaniel is this something you plan to implement? Or should I do it?

**Comment by zachdaniel** [comment]
I was not planning on doing it.
